### PR TITLE
fix: mapreduce type stability, VoA broadcast adjoints

### LINKED ISSUE #323: Type instability when iterating nested recursive arrays
I noticed that the following is type unstable (on v3.3.3 and julia 1.10.0)

```julia
using RecursiveArrayTools

sum(ArrayPartition(ArrayPartition(zeros(4,4)))) # ::Any

sum(VectorOfArray([zeros(4), VectorOfArray([zeros(4,4)])])) # ::Any
```

Curiously, all of the following are stable.
```julia
sum(ArrayPartition(ArrayPartition(zeros(4)))) # ::Float64
sum(ArrayPartition(zeros(4,4))) # ::Float64

sum(VectorOfArray([VectorOfArray([zeros(4,4)])])) # ::Float64
```

The other recursive array types seem to be affected as well. Not sure if getting these stable is in-scope, but it would be super helpful.

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Close #323 

## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

### CODE REVIEW

File: src/vector_of_array.jl [hunk 1]
```diff
@@ -595,8 +604,14 @@ function Base.convert(::Type{Array}, VA::AbstractVectorOfArray)
 end
 
 # statistics
-@inline Base.sum(f, VA::AbstractVectorOfArray) = sum(f, Array(VA))
-@inline Base.sum(VA::AbstractVectorOfArray; kwargs...) = sum(Array(VA); kwargs...)
+@inline Base.sum(VA::AbstractVectorOfArray; kwargs...) = sum(identity, VA; kwargs...)
+@inline function Base.sum(f, VA::AbstractVectorOfArray; kwargs...)
+    if hasproperty(kwargs, :dims)
+        sum(Array(VA); kwargs...)
```
Comment (ChrisRackauckas) [inline src/vector_of_array.jl:610]: This can be a potentially big allocation

File: src/vector_of_array.jl [hunk 2]
```diff
@@ -638,8 +653,9 @@ end
 end
 
 Base.map(f, A::RecursiveArrayTools.AbstractVectorOfArray) = map(f, A.u)
+
 function Base.mapreduce(f, op, A::AbstractVectorOfArray)
```
Comment (ChrisRackauckas) [inline src/vector_of_array.jl:657]: probably need to have an `init`

File: src/vector_of_array.jl (hunk 1 above)
Comment (AayushSabharwal) [reply src/vector_of_array.jl:614]: That was the default implementation, but I just realized two things:
- that method is probably not necessary
- mapreduce is implemented wrong

Comment (AayushSabharwal) [reply src/vector_of_array.jl]: Fixed in https://github.com/SciML/RecursiveArrayTools.jl/pull/325/commits/bdb58305372db57d38877e48157859e3137f7ccc

File: src/vector_of_array.jl (hunk 2 above)
Comment (AayushSabharwal) [reply src/vector_of_array.jl]: Fixed in https://github.com/SciML/RecursiveArrayTools.jl/pull/325/commits/bdb58305372db57d38877e48157859e3137f7ccc

File: ext/RecursiveArrayToolsZygoteExt.jl [hunk 3]
```diff
@@ -117,19 +122,122 @@ end
...
+    broadcast(+, x, y), ȳ -> (nothing, map(x -> Zygote.unbroadcast(x, ȳ), (x, y))...)
+end
+@adjoint function Broadcast.broadcasted(::typeof(+), x::Zygote.Numeric, y::AbstractVectorOfArray)
+    broadcast(+, x, y), ȳ -> (nothing, map(x -> Zygote.unbroadcast(x, ȳ), (x, y))...)
+end
+
+_minus(Δ) = .-Δ
+_minus(::Nothing) = nothing
+
+@adjoint function Broadcast.broadcasted(::typeof(-), x::AbstractVectorOfArray, y::Union{AbstractVectorOfArray, Zygote.Numeric})
```
Comment (ChrisRackauckas) [inline ext/RecursiveArrayToolsZygoteExt.jl:152]: you're never supposed to define specific broadcasts like this, and I don't think digging into Zygote's broadcast system is the answer. It just relies on the Julia-level broadcast

**Comment by ChrisRackauckas** [comment]
This will need a real fix, but let's at least get downstream tests passing.

**Pull request merged**